Internal #2577: Window Atomic Finalize

hawkfish opened the pull request:
The finalized counter now needs to be an atomic
because it is not always protected by a mutex.

Commit:
Internal #2577: Window Atomic Finalize

The finalized counter now needs to be an atomic
because it is not always protected by a mutex.

carlopi reviewed (approved):
Change is safe, and looks to solve the Thread Sanitizer reported problem.

Question on "it is not always protected by a mutex", should it be? That means there is a problem elsewhere, or it's actual the proper solution to use atomic?

Apart from curiosity, this looks good. Thanks

---

Failure is unrelated.

hawkfish commented:
> Question on "it is not always protected by a mutex", should it be? That means there is a problem elsewhere, or it's actual the proper solution to use atomic?

It is the proper solution. Atomics are basically counters with building mutexes, but they are lighter weight. You can increment them and get the result of the increment without a data race.

Commit:
Merge branch 'window-finalized' into window-distinct

Commit:
Internal #2577: Window Tree Allocation

The ArenaAllocators used to build the segment tree in parallel
need to have lifetimes tied to the tree, not the local state that
was used to construct it.

fixes: duckdblabs/duckdb-internal#2577

Commit:
Merge branch 'window-arena' into window-distinct

Commit:
Merge branch 'main' into window-finalized

Mytherin commented:
Thanks!
